common/rsyslog_plugin.cpp

```diff
@@ -0,0 +1,161 @@
+#include <iostream>
+#include <vector>
+#include <string>
+#include <cstdlib>
+#include <cstring>
+#include <fstream>
+#include <unordered_set>
+#include <regex>
+#include <tuple>
+#include "logger.h"
+#include "json.hpp"
+
+using namespace std;
+using namespace nlohmann;
+
+string parserCfgFile = "{}_parse.rc.json";
+ofstream outfile;
+string logfile = "/tmp/{}_logfile";
+json regexList = json::array();
+string pgmName = "";
+unordered_set<string> errReportedTags;
+
+void loadRegexlist(string rcPath, string pgmName) {
+	fstream rcfile;
+	parserCfgFile = regex_replace(parserCfgFile, regex("\\{}"), pgmName);
+	parserCfgFile = rcPath + parserCfgFile;
+	rcfile.open(parserCfgFile, ios::in);
+	if (!rcfile) {
+		cerr << "No such file: " << parserCfgFile << endl;
+		return;
+	}
+	rcfile >> regexList;
+}
+
+tuple<bool, json> matchRegex(string msg) {
+	bool isMatch = false;
+	json structuredEvent;
+	for (int i = 0; i < regexList.size(); i++) {
+		string regexString = regexList[i]["regex"];
+		regex expression(regexString);
```

Comment:
>regex expression(regexString); [](http://example.com/codeflow?start=2&length=30)

You repeatedly construct the same list of regex per msg, which is not efficient. The list of regex is never changed and you should reuse.

To match a list of regex, you may combine them into one combo regex. Seem you may use `regex_iterator` to write an high performance search function.

Reply:
I don't think we can combine into one combo regex, because each regex will correspond to different set of parameters.
For example, taking regex list of [x,y,z] and parameter list of [[x1,x2],[y1],[z1,z2]], if we were to combine into one regex such that we get (x)|(y)|(z), and our message matches one of the regexes, we won't know what parameter list to use for our capture group.

Reply:
"combine into one combo regex" is a low priority discussion.

How about other suggestions in this comment.

Reply:
With new commit, I avoid creating regex list per message